docs/constrained-generators/manual.md

```diff
@@ -324,125 +568,1080 @@ match
+the solution for `a`, and then constructing a `(Term b)` obtained from this value.
 
-# Predefined HasSpec instances.
+```
+ex7 :: Specification (Int,[Int])
+ex7 = constrained $ \ pair ->
+      match pair $ \ n xs ->
+      reifies n xs sum
+```
 
-In order to write specification for a particular type, that type must have a `HasSpec` instance. 
-A type with a `HasSpec` instance might have a number of Function Symbols that operate on that type.
-There are a number of types that have predefined `HasSpec` instances. We list them here along with the
-type of their function symbols.
+Note that `sum` is a Haskell function, not a function symbol in the system that can be applied to `Term`s.
+The system solves for `xs`,
+then  applies `sum` to the list obtained by solving `xs`. Then binds the `Term` variable `n` to that literal value.
+Here is a sample solution.
 
-## Numeric types
+```
+ghci> debugSpec ex7
+constrained $ \ v_3 ->
+  [to v_3/v_2]let v_1 = prodFst_ v_2 in
+              let v_0 = prodSnd_ v_2 in reifies v_1 v_0
+(82,[-24,22,28,19,16,24,-20,1,16])
+True
+```
 
-`(Int,Integer, Natural, Int8, Int16, Int32, Int64, Word8, Word16, Word32, Word62)`
+The second operation `reify` can be defined using `reifies` by placing an existential constraint on the range of the function.
+Here is an example of the use of `reify`
 
- 1. `(<=.) :: OrdLike a => Term a -> Term a -> Term Bool`
- 2. `(<.) :: OrdLike a => Term a -> Term a -> Term Bool`
- 3. `(>=.) :: OrdLike a => Term a -> Term a -> Term Bool`
- 4. `(>.) :: OrdLike a => Term a -> Term a -> Term Bool`
- 5. `(==.) :: HasSpec a => Term a -> Term a -> Term Bool`
- 6.  Num instance for (Term n) where n is a Numeric type. Operators `(+)`, `(-)`, `(*)`
- 
-## `HasSpec Bool`
+```
+ex8 :: Specification ([Int],[Int])
+ex8 = constrained $ \ pair ->
+      match pair $ \ xs1 xs2 -> 
+       [ assert $ sizeOf_ xs1 <=. 5
+       , forAll xs1 $ \ x ->  x <=. 10
+       , reify xs1 reverse $ \ t -> xs2 ==. t
+       ]
+```
 
-  1. `or_ :: Term Bool -> Term Bool -> Term Bool`
-  2.  `not_ :: Term Bool -> Term Bool`
+Here we are saying there exists `t` which can be obtained by `reverse`ing `xs1` and that `xs2` equals `t`
+Here is a sample solution.
+
+```
+ghci> debugSpec ex8
+constrained $ \ v_4 ->
+  [to v_4/v_3]let v_2 = prodFst_ v_3 in
+              let v_1 = prodSnd_ v_3 in
+              {assert $ sizeOf_ v_2 <=. 5
+               forall v_0 in v_2 $ assert $ v_0 <=. 10
+               exists v_0 in
+               {reifies v_0 v_2
+                Explain ["reify v_2 somef $"] $ assert $ v_1 ==. v_0}}
+([9,-1,-11],[-11,-1,9])
+True
+```
+
+The third operation `assertReified` can be used to place a boolean constraint on the existential.
+Here is an example of its use.
+
+```
+ex9 :: Specification Int
+ex9 = constrained $ \x ->
+  [ assert $ x <=. 10
+  , assertReified x (<= 3) -- Note that (<= 3) is a Bool function on values, not Terms.
+  ]
+```
```

Comment:
It might be better to use an example with e.g. a `forAll` or a `caseOn` for this as `assertReified` is basically deferring to `suchThat`. In fact, that's probably something we should be pointing out.

Reply:
done